src/parse/parse-token-request.ts

```diff
@@ -0,0 +1,299 @@
+      ))
+  ) {
+    //We need to error here to match the validation in the token handler
+    //This matches the validation here https://github.com/govuk-one-login/authentication-api/blob/dc42596ab02184f80c30f327a7dcd0f76146e619/orchestration-shared/src/main/java/uk/gov/di/orchestration/shared/services/ClientSignatureValidationService.java#L103 where the audience is validated in the client signature
+    // validation service
+    logger.warn(
+      `Invalid audience in client JWT assertion: Expected: ${EXPECTED_PRIVATE_KEY_JWT_AUDIENCE}, Received: ${parsedClientAssertion.payload.aud}`
+    );
+
+    throw new TokenRequestError({
+      errorCode: "invalid_client",
+      errorDescription: "Invalid signature in private_key_jwt",
+      httpStatusCode: 400,
+    });
+  }
+
+  if (parsedClientAssertion.payload.sub !== parsedClientAssertion.payload.iss) {
+    logger.warn(
+      {
+        iss_claim: parsedClientAssertion.payload.iss,
+        sub_claim: parsedClientAssertion.payload.sub,
+      },
+      "client_assertion iss and sub claims do not match"
+    );
+
+    throw new TokenRequestError({
+      errorCode: "invalid_client",
+      errorDescription: "Invalid signature in private_key_jwt",
+      httpStatusCode: 400,
+    });
+  }
+
+  if (
+    !(await isSignatureValid(
+      clientPublicKey,
+      tokenSigningAlgorithm,
+      parsedClientAssertion.token
+    ))
+  ) {
+    logger.warn("Could not verify signature of private_key_jwt");
+    throw new TokenRequestError({
+      errorCode: "invalid_client",
+      errorDescription: "Invalid signature in private_key_jwt",
+      httpStatusCode: 400,
+    });
+  }
+
+  return {
+    tokenRequest: {
+      grant_type: tokenRequestBody.grant_type,
+      redirectUri: tokenRequestBody.redirect_uri,
+      code: tokenRequestBody.code,
+      client_assertion: tokenRequestBody.client_assertion,
+      client_assertion_type: tokenRequestBody.client_assertion_type,
+    },
+    clientAssertion: parsedClientAssertion,
+  };
+};
+
+const parseClientAssertion = (
+  clientAssertion: string,
+  tokenRequestClientId?: string
+): PrivateKeyJwt => {
+  const jwtParts = clientAssertion.split(".");
+
+  if (jwtParts.length !== 3) {
+    throw new ParseRequestError(
+      "Invalid client_assertion JWT: Unexpected number of Base64URL parts, must be three"
+    );
+  }
+
+  const payload = decodeJwt(clientAssertion);
+  const header = decodeProtectedHeader(clientAssertion);
+  const signature = jwtParts[2];
+
+  if (!signature || signature.trim().length === 0) {
+    throw new ParseRequestError(
+      "Invalid client_assertion JWT: The signature must not be empty"
+    );
+  }
+
+  if (!payload.sub) {
+    throw new ParseRequestError(
+      "Invalid client_assertion JWT: Missing subject in client JWT assertion"
+    );
+  }
+
+  if (!payload.iss) {
+    throw new ParseRequestError(
+      "Invalid client_assertion JWT: Missing issuer in client JWT assertion"
+    );
+  }
+
```

Comment:
Right here, I think we should have a valid JSON check, with `Payload of JWS object is not a valid JSON object` from `getJWTClaimsSet`

Reply:
Looking at the source code for jose, https://github.com/panva/jose/blob/4f471eb3ec64c1a1c58d5fa211eaeb87fa8d39e0/src/util/decode_jwt.ts#L46, when we call decodeJWT it will throw an error for invalid JSON. This should align with the real token endpoint as before we check iss or sub claims we need to call `getJWTClaimsSet`. I think I should catch this InvalidJWT error and rethrow it as a Parse error to ensure we capture it correctly

Reply:
Have done this now!